PhysicsTools/PatAlgos/plugins/CandidateSummaryTable.cc

```diff
@@ -12,65 +12,87 @@
-    public:
-      explicit CandidateSummaryTable(const edm::ParameterSet & iConfig);
-      ~CandidateSummaryTable();
-
-      virtual void analyze(const edm::Event & iEvent, const edm::EventSetup& iSetup) override;
-      virtual void endJob() override;
-
-    private:
-      struct Record {
-        edm::InputTag src;
-        edm::EDGetTokenT<edm::View<reco::Candidate> > srcToken;
-        size_t present, empty, min, max, total;
-        Record(edm::InputTag tag, edm::ConsumesCollector && iC) : src(tag), srcToken(iC.consumes<edm::View<reco::Candidate> >(tag)), present(0), empty(0), min(0), max(0), total(0) {}
-
-        void update(const edm::View<reco::Candidate> &items) {
-            present++;
-            size_t size = items.size();
-            if (size == 0) {
-                empty++;
-            } else  {
-                if (min > size) min = size;
-                if (max < size) max = size;
-            }
-            total += size;
-        }
-      };
-      std::vector<Record> collections_;
-      size_t totalEvents_;
-      bool perEvent_, perJob_;
-      std::string self_, logName_;
-      bool dumpItems_;
+namespace pathelpers {
+  struct Record {
+    const edm::InputTag src;
+    std::atomic<size_t> present, empty, min, max, total;
+    Record(const edm::InputTag& tag) : src(tag), present(0), empty(0), min(0), max(0), total(0) {}
+    Record(const Record& other) : src(other.src), 
+                                  present(other.present.load()), 
+                                  empty(other.empty.load()), 
+                                  min(other.min.load()), 
+                                  max(other.max.load()), 
+                                  total(other.total.load()) {}
+    void update(const edm::View<reco::Candidate> &items) {
+      present++;
+      const size_t size = items.size();
+      if (size == 0) {
+        empty++;
+      } else  {
+        if (min > size) min = size;
+        if (max < size) max = size;
+      }
+      total += size;
+    }
   };
 
+  struct RecordCache { 
+    RecordCache(const edm::ParameterSet& iConfig) : 
+      perEvent_(iConfig.getUntrackedParameter<bool>("perEvent", false)),
+      perJob_(iConfig.getUntrackedParameter<bool>("perJob", true)),
+      self_(iConfig.getParameter<std::string>("@module_label")),
+      logName_(iConfig.getUntrackedParameter<std::string>("logName")),
+      dumpItems_(iConfig.getUntrackedParameter<bool>("dumpItems", false)),
+      totalEvents_(0) {
+      const std::vector<edm::InputTag>& tags  = iConfig.getParameter<std::vector<edm::InputTag> >("candidates");
+      for( const auto& tag : tags ) {
+        collections_.emplace_back( tag );
+      }
+    }
+    const bool perEvent_, perJob_;
+    const std::string self_, logName_;
+    const bool dumpItems_;
+    mutable std::atomic<size_t> totalEvents_;
+    mutable std::vector<Record> collections_; // size of vector is never altered later! (atomics are in the class below)
```

Comment:
You should make this non-mutable and have the thread safe parts of Record mutable.